File: dotnet/System/Workspace/Adapters/Allors.Workspace.Adapters.Local/Database/DatabaseConnection.cs
// <copyright file="v.cs" company="Allors bv">
// Copyright (c) Allors bv. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Local
{
    using System;
    using System.Collections.Concurrent;
    using System.Collections.Generic;
    using System.Linq;
    using Database;
    using Database.Security;
    using Database.Services;
    using Meta;
    using Ranges;
    using AccessControl = AccessControl;
    using IRoleType = Database.Meta.IRoleType;

    public class DatabaseConnection : Adapters.DatabaseConnection
    {
        private readonly Dictionary<long, AccessControl> accessControlById;
        private readonly ConcurrentDictionary<long, DatabaseRecord> recordsById;

        private readonly Func<IWorkspaceServices> servicesBuilder;
        private readonly IRanges<long> recordRanges;

        public DatabaseConnection(Configuration configuration, IDatabase database, Func<IWorkspaceServices> servicesBuilder, Func<IRanges<long>> rangesFactory) : base(configuration, new IdGenerator())
        {
            this.Database = database;
            this.servicesBuilder = servicesBuilder;
            this.recordRanges = rangesFactory();

            this.recordsById = new ConcurrentDictionary<long, DatabaseRecord>();
            this.accessControlById = new Dictionary<long, AccessControl>();
        }

        public long UserId { get; set; }

        private IDatabase Database { get; }

        public IDatabaseServices DatabaseServices => this.Database.Services;

        public Database.Meta.IMetaPopulation MetaPopulation => this.Database.MetaPopulation;

        public ITransaction CreateTransaction()
        {
            var transaction = this.Database.CreateTransaction();
            var user = (IUser)transaction.Instantiate(this.UserId);
            transaction.Services.Get<IUserService>().User = user;
            return transaction;
        }

        internal void Sync(IEnumerable<IObject> objects, IAccessControl accessControl)
        {
            using (var transaction = this.Database.CreateTransaction())
            {
                foreach (var @object in objects)
                {
                    var id = @object.Id;
                    var databaseClass = @object.Strategy.Class;
                    var roleTypes = databaseClass.DatabaseRoleTypes.Where(w => w.RelationType.WorkspaceNames.Length > 0);

                    var workspaceClass = (IClass)this.Configuration.MetaPopulation.FindByTag(databaseClass.Tag);
                    var roleByRoleType = roleTypes.ToDictionary(w => ((IRelationType)this.Configuration.MetaPopulation.FindByTag(w.RelationType.Tag)).RoleType, w => this.GetRole(@object, w));

                    var acl = accessControl[@object];

                    var accessControls = acl.Grants?.Select(v => (IGrant)transaction.Instantiate(v.Id)).Select(this.GetAccessControl).ToArray() ?? Array.Empty<AccessControl>();

                    this.recordsById[id] = new DatabaseRecord(workspaceClass, id, @object.Strategy.ObjectVersion, roleByRoleType, this.recordRanges.Load(acl.Revocations.Select(v => v.Id)), accessControls);
                }
            }
        }

        public override IWorkspace CreateWorkspace() => new Workspace(this, this.servicesBuilder(), this.recordRanges);

        public override Adapters.DatabaseRecord GetRecord(long id)
        {
            this.recordsById.TryGetValue(id, out var databaseObjects);
            return databaseObjects;
        }

        public override long GetPermission(IClass workspaceClass, IOperandType operandType, Operations operation)
        {
            var @class = (Database.Meta.IClass)this.Database.MetaPopulation.FindByTag(workspaceClass.Tag);
            var operandId = this.Database.MetaPopulation.FindByTag(operandType.OperandTag).Id;

            long permission;
            switch (operation)
            {
                case Operations.Read:
                    @class.ReadPermissionIdByRelationTypeId.TryGetValue(operandId, out permission);
                    break;
                case Operations.Write:
                    @class.WritePermissionIdByRelationTypeId.TryGetValue(operandId, out permission);
                    break;
                case Operations.Execute:
                    @class.ExecutePermissionIdByMethodTypeId.TryGetValue(operandId, out permission);
                    break;
                case Operations.Create:
                    throw new NotSupportedException("Create is not supported");
                default:
                    throw new ArgumentOutOfRangeException($"Unknown operation {operation}");
            }

            return permission;
        }

        internal IEnumerable<IObject> ObjectsToSync(Pull pull) =>
            pull.DatabaseObjects.Where(v =>
            {
                if (this.recordsById.TryGetValue(v.Id, out var databaseRoles))
                {
                    return v.Strategy.ObjectVersion != databaseRoles.Version;
                }

                return true;
            });

        private AccessControl GetAccessControl(IGrant grant)
        {
            if (!this.accessControlById.TryGetValue(grant.Strategy.ObjectId, out var acessControl))
            {
                acessControl = new AccessControl();
                this.accessControlById.Add(grant.Strategy.ObjectId, acessControl);
            }

            if (acessControl.Version == grant.Strategy.ObjectVersion)
            {
                return acessControl;
            }

            acessControl.Version = grant.Strategy.ObjectVersion;
            acessControl.PermissionIds = this.recordRanges.Import(grant.Permissions.Select(v => v.Id));

            return acessControl;
        }

        private object GetRole(IObject @object, IRoleType roleType)
        {
            if (roleType.ObjectType.IsUnit)
            {
                return @object.Strategy.GetUnitRole(roleType);
            }

            if (roleType.IsOne)
            {
                return @object.Strategy.GetCompositeRole(roleType)?.Id;
            }

            return this.recordRanges.Load(@object.Strategy.GetCompositesRole<IObject>(roleType).Select(v => v.Id));
        }


    }
}

File: typescript/libs/system/workspace/adapters-json/src/lib/database/database-connection.ts
import {
  PullResponse,
  SyncRequest,
  SyncResponse,
  AccessRequest,
  AccessResponse,
  PermissionRequest,
  PermissionResponse,
} from '@allors/system/common/protocol-json';
import {
  Grant,
  DatabaseConnection as SystemDatabaseConnection,
  MapMap,
  Revocation,
} from '@allors/system/workspace/adapters';
import {
  Configuration,
  IWorkspace,
  Operations,
} from '@allors/system/workspace/domain';
import {
  Class,
  MethodType,
  OperandType,
  RelationType,
} from '@allors/system/workspace/meta';
import { DatabaseRecord } from './database-record';
import { ResponseContext } from './security/response-context';
import { Workspace } from '../workspace/workspace';
import { IDatabaseJsonClient } from '../idatabase-json-client';

export class DatabaseConnection extends SystemDatabaseConnection {
  private recordsById: Map<number, DatabaseRecord>;

  grantById: Map<number, Grant>;
  revocationById: Map<number, Revocation>;
  permissions: Set<number>;

  readPermissionByOperandTypeByClass: MapMap<Class, OperandType, number>;
  writePermissionByOperandTypeByClass: MapMap<Class, OperandType, number>;
  executePermissionByOperandTypeByClass: MapMap<Class, OperandType, number>;

  constructor(
    configuration: Configuration,
    public client: IDatabaseJsonClient
  ) {
    super(configuration);

    this.recordsById = new Map();

    this.grantById = new Map();
    this.revocationById = new Map();
    this.permissions = new Set();

    this.readPermissionByOperandTypeByClass = new MapMap();
    this.writePermissionByOperandTypeByClass = new MapMap();
    this.executePermissionByOperandTypeByClass = new MapMap();
  }

  createWorkspace(): IWorkspace {
    return new Workspace(this);
  }

  onPullResonse(response: PullResponse, context: string): SyncRequest {
    return {
      x: context,
      o: response.p
        .filter((v) => {
          const record = this.recordsById.get(v.i);

          if (record == null) {
            return true;
          }

          if (record.version !== v.v) {
            return true;
          }

          if (!this.ranges.equals(record.grants, v.g)) {
            return true;
          }

          if (!this.ranges.equals(record.revocations, v.r)) {
            return true;
          }

          return false;
        })
        .map((v) => v.i),
    };
  }

  onSyncResponse(syncResponse: SyncResponse): AccessRequest | null {
    const ctx = new ResponseContext(this);

    for (const syncResponseObject of syncResponse.o) {
      const databaseObjects = DatabaseRecord.fromResponse(
        this,
        ctx,
        syncResponseObject
      );
      this.recordsById.set(databaseObjects.id, databaseObjects);
    }

    if (ctx.missingGrantIds.size > 0 || ctx.missingRevocationIds.size > 0) {
      return {
        g: Array.from(ctx.missingGrantIds),
        r: Array.from(ctx.missingRevocationIds),
      };
    }

    return null;
  }

  accessResponse(
    accessResponse: AccessResponse
  ): PermissionRequest | undefined {
    let missingPermissionIds: Set<number> | undefined = undefined;

    if (accessResponse.g != null) {
      for (const accessResponseGrant of accessResponse.g) {
        const id = accessResponseGrant.i;
        const version = accessResponseGrant.v;
        const permissionIds = accessResponseGrant.p;
        this.grantById.set(id, new Grant(version, permissionIds));

        if (permissionIds != null) {
          for (const permissionId of permissionIds) {
            if (this.permissions.has(permissionId)) {
              continue;
            }

            missingPermissionIds ??= new Set();
            missingPermissionIds.add(permissionId);
          }
        }
      }
    }

    if (accessResponse.r != null) {
      for (const accessResponseRevocation of accessResponse.r) {
        const id = accessResponseRevocation.i;
        const version = accessResponseRevocation.v;
        const permissionIds = accessResponseRevocation.p;
        this.revocationById.set(id, new Revocation(version, permissionIds));

        if (permissionIds != null) {
          for (const permissionId of permissionIds) {
            if (this.permissions.has(permissionId)) {
              continue;
            }

            missingPermissionIds ??= new Set();
            missingPermissionIds.add(permissionId);
          }
        }
      }
    }

    if (missingPermissionIds != null) {
      return {
        p: Array.from(missingPermissionIds),
      };
    }

    return undefined;
  }

  permissionResponse(permissionResponse: PermissionResponse): void {
    if (permissionResponse.p != null) {
      for (const syncResponsePermission of permissionResponse.p) {
        const id = syncResponsePermission.i;
        const cls = this.configuration.metaPopulation.metaObjectByTag.get(
          syncResponsePermission.c
        ) as Class;
        const metaObject =
          this.configuration.metaPopulation.metaObjectByTag.get(
            syncResponsePermission.t
          );
        const operandType: OperandType =
          (metaObject as RelationType)?.roleType ?? (metaObject as MethodType);
        const operation = syncResponsePermission.o;

        this.permissions.add(id);

        switch (operation) {
          case Operations.Read:
            this.readPermissionByOperandTypeByClass.set(cls, operandType, id);
            break;
          case Operations.Write:
            this.writePermissionByOperandTypeByClass.set(cls, operandType, id);
            break;
          case Operations.Execute:
            this.executePermissionByOperandTypeByClass.set(
              cls,
              operandType,
              id
            );
            break;
          case Operations.Create:
            throw new Error('Create is not supported');
          default:
            throw new Error('Argument out of range');
        }
      }
    }
  }

  getPermission(
    cls: Class,
    operandType: OperandType,
    operation: Operations
  ): number | undefined {
    switch (operation) {
      case Operations.Read:
        return this.readPermissionByOperandTypeByClass.get(cls, operandType);
      case Operations.Write:
        return this.writePermissionByOperandTypeByClass.get(cls, operandType);
      case Operations.Execute:
        return this.executePermissionByOperandTypeByClass.get(cls, operandType);
      case Operations.Create:
        throw new Error('Create is not supported');
      default:
        throw new Error('Argument out of range');
    }
  }

  getRecord(id: number): DatabaseRecord | undefined {
    return this.recordsById.get(id);
  }
}
